### PR TITLE
ARROW-11705: [R] Support scalar value recycling in RecordBatch/Table$create()

### PR DESCRIPTION
**Pull request by thisisnic**
This also adds missing spaces in some unrelated R files

### CODE REVIEW

File: r/R/table.R
Comment (ianmcook) [inline r/R/table.R:181]: In `arrow-package.R`, add `modify2` to the line beginning with `#' @importFrom purrr`. Then do `devtools::document()` to update the `NAMESPACE` file. Then remove the `purrr::` from this line.

**Comment by thisisnic** [comment]
@ianmcook I think this is something for a separate ticket/PR, but when I was testing things you mentioned above, I found that it is possible to create `Table` and `RecordBatch` objects with duplicated column names, which then results in errors when I try to analyse them, e.g. 

```r
Table$create(iris, iris) %>% filter(Species == "versicolor")
```
` Error in schm$GetFieldByName(name)$ToString() : attempt to apply non-function `

Shall I create a new ticket to add code which checks column names are unique when combining objects?

File: r/R/record-batch.R
Comment (ianmcook) [inline r/R/record-batch.R:168]:
```suggestion
  arr_lens <- map_int(arrays, length)
  if (length(arrays) > 1 && any(arr_lens == 1) && !all(arr_lens==1)){
    max_array_len <- max(arr_lens)
```
(and the same in `table.R`)

File: r/R/table.R
Comment (ianmcook) [inline r/R/table.R:181]:
```suggestion
  arr_lens <- map_int(dots, length)
  if (length(dots) > 1 && any(arr_lens == 1) && !all(arr_lens==1)){
    max_array_len <- max(arr_lens)
```

Comment (thisisnic) [reply r/R/table.R:181]: Oops, must have escaped as I was moving things around!

File: r/R/record-batch.R
Comment (ianmcook) [inline r/R/record-batch.R:172]: We should find a way to do this without calling `as.vector()` because that will convert Arrow objects to R vectors which could cause the data type to be lost during the conversion.

But the problem is that without `as.vector()`, length 1 `ChunkedArray` objects will error when passed to `Scalar$create()`. I think the cleanest way to solve that is to improve the `Array$create` function in `array.R` to handle `ChunkedArray` objects and convert it to `Array`. (This is getting into yak shaving territory here but I think it's important.)

Comment (ianmcook) [reply r/R/record-batch.R:172]: Looking at this a bit more, I think it might require some challenging C++ coding to implement `combine_chunks()` like I described. Perhaps it should be dealt with in a separate Jira & separate PR.

Comment (thisisnic) [reply r/R/record-batch.R]: Done!

**Comment by thisisnic** [comment]
> This looks pretty good to me! Just a few final things:
> 
>     * Could you please ensure there are spaces added: `if(` → `if (` and `){`→ `) {`
> 
>     * Could you search all the tests for "[ARROW-11705](https://issues.apache.org/jira/browse/ARROW-11705)" and see if the two skipped tests work now that you've implemented this fix?

Done both now, and fixed spacing in a load of other places as well.

File: r/R/record-batch.R
Comment (ianmcook) [reply r/R/record-batch.R]: Thanks! I created ARROW-12737 for follow up on `concat_arrays()` and `combine_chunks()`

**Review by ianmcook — APPROVED** [review]

**Comment by ianmcook** [comment]
@nealrichardson you want to take a look before I merge? Thanks

**Comment by ianmcook** [comment]
@thisisnic please also build the docs so `util.Rd` is included here

**Comment by ianmcook** [comment]
In theory, this would be a good opportunity to use Conbench to test whether the multiple `length()` calls added here might have a meaningful effect on the performance of `Table`/`RecordBatch` creation. In practice, I'm not sure whether Conbench would help us here. @jonkeane do you know?

**Comment by jonkeane** [comment]
It's worth checking — there might be too much noise right now to see a small difference, but large ones should jump out.

**Comment by jonkeane** [comment]
@ursabot please benchmark lang=R

**Comment by thisisnic** [comment]
> Benchmark runs are scheduled for baseline = [4e0f0cf](https://github.com/apache/arrow/commit/4e0f0cf79cf836a29e4bfd4a7b2d692f8b50bffe) and contender = [dbe7491](https://github.com/apache/arrow/commit/dbe74918019e172f8bdd3a2085f1ec7481fa79f4). Results will be available as each benchmark for each run completes.
> Conbench compare runs links:
> [Failed arrow_down0.0% arrow_up0.0%] [ec2-t3-large-us-east-2 (mimalloc)](https://conbench.ursa.dev/compare/runs/614f005c2822403591fddd25921d2358...67131004c3734086a26def2729031134/)
> [Failed arrow_down0.0% arrow_up0.0%] [ec2-t3-xlarge-us-east-2 (mimalloc)](https://conbench.ursa.dev/compare/runs/c61a9f2f9b504ad49587818f3ae5f283...0e8346acd8a040f9bdbfe92576c027f5/)
> [Scheduled] [ursa-i9-9960x (mimalloc)](https://conbench.ursa.dev/compare/runs/139601ccb33e49c88e039fcfa1a6d460...1a688615902f4a0a997f48e908738a25/)
> [Failed arrow_down0.0% arrow_up0.0% Warning: Contender and baseline run contexts do not match] [ursa-thinkcentre-m75q (mimalloc)](https://conbench.ursa.dev/compare/runs/ab3cde1618b84feea09029227d19e223...59d9724c63334b16ba4b5d1f20e20875/)
> warning ursa-i9-9960x agent is disconnected or machine is offline.

@jonkeane It looks like Conbench has used different benchmarks for each of those PRs - do you know why that's happened?

**Comment by ianmcook** [comment]
I think Conbench needs some tweaking before it can help us here.

I'll go ahead and resolve the conflicts, wait for checks to pass, and merge this if there aren't any objections

**Comment by ianmcook** [comment]
@thisisnic ooh we finally have some benchmark results to look at!: [ursa-i9-9960x (mimalloc)](https://conbench.ursa.dev/compare/runs/139601ccb33e49c88e039fcfa1a6d460...1a688615902f4a0a997f48e908738a25/) 
The `dataframe-to-table` results are the pertinent ones here I think.
@jonkeane can you help us interpret these results?

**Comment by thisisnic** [comment]
> @thisisnic ooh we finally have some benchmark results to look at!: [ursa-i9-9960x (mimalloc)](https://conbench.ursa.dev/compare/runs/139601ccb33e49c88e039fcfa1a6d460...1a688615902f4a0a997f48e908738a25/)
> The `dataframe-to-table` results are the pertinent ones here I think.
> @jonkeane can you help us interpret these results?

I'm going to take a stab at interpreting them just to see how I do.  Overall, the changes I've made have made things 5% slower.  I'm not sure whether this is important - as I don't have a good idea of a cut-off, or idea whether any of this is just noise.  Looking more closely at the results, the biggest differences are to file-read, where the update is at least 8% slower.

Does this feel like a significant slowdown? I'd say so.  I think it provides support to the idea that this really should be implemented at the C++ level rather than the R level.  Is this OK to be merged in? I think "perhaps", as long as we open a JIRA for this to be implemented at the C++ level.

Let me know your thoughts!

**Comment by jonkeane** [comment]
A couple of comments/additions, that I think you're generally right.

The R benchmarks tend to be stable (https://conbench.ursa.dev/compare/runs/8b6fef07829948998502a7677dec6e03...0cbd9dcbe2594e06ab95cf0e088cf25b/ is a run on the master branch and is between -3% and 1% change and that -3% is an outlier there, the next largest decrease is -0.8%). So we can have decent confidence that we're not observing noise alone here. We're working actively to improve this, but wanted to put it out there as part of the assumptions I'm using.

There are some file-read benchmarks that are >5% slower, interestingly it is all (and only) the fanniemae dataset that is slower (both reading from parquet and from feather) and *only* when it is being converted to a data.frame, not when it is being left as a table. This seems a little suspect to me since the only places that I'm seeing you've meaningfully changed the code is `RecordBatch$create`, `Table$create`, and `MakeArrayFromScalar`. Do any of those get called when reading parquet or feather files? 

Note: I don't see csv reads run here, IIRC those were proactively disabled due to memory issues, but I will confirm that (and I thought this machine should have been able to handle these and there is https://issues.apache.org/jira/browse/ARROW-12519 to track).

There are also another number of benchmarks that are in the 5-1% slower range (the other file-read, as well as the df to R conversions, and a handful of the writing benchmarks). The df to R conversions seem more in line with the code that was changed, and those are in the 3-6% range (though most are closer to 3%, with one being an outlier at 6%)

The next 28/128 or ~20% of the benchmarks are 0-1% slower and then 19/138 or ~14% of the benchmarks are 0-1% faster. These are probably all just noise.

**Comment by thisisnic** [comment]
> There are some file-read benchmarks that are >5% slower, interestingly it is all (and only) the fanniemae dataset that is slower (both reading from parquet and from feather) and _only_ when it is being converted to a data.frame, not when it is being left as a table. This seems a little suspect to me since the only places that I'm seeing you've meaningfully changed the code is `RecordBatch$create`, `Table$create`, and `MakeArrayFromScalar`. Do any of those get called when reading parquet or feather files?

They do not, which does make it strange; completely overlooked the fact that those shouldn't be relevant here.

File: r/R/scalar.R
Comment (nealrichardson) [inline r/R/scalar.R:36]: If you pass this through, maybe you don't need `repeat_value_as_array`

```suggestion
    as_array = function(length = 1L) MakeArrayFromScalar(self, as.integer(length)),
```

File: r/R/record-batch.R
Comment (nealrichardson) [inline r/R/record-batch.R:165]: This code block seems to be repeated in both the Table and RecordBatch code, so it might be better factored out as a helper function.

I also wonder whether the logic would actually be simpler in C++ because you could do it at a later point where you know exactly that you have a vector of Arrays and don't have to worry about whether it is an R vector, a Scalar, an Array, etc. See the `check_consistent_array_size` function for example in `r/src`--you could drop in around there and instead of erroring if you don't have consistent lengths, handle the recycling case. (Also ok to (1) ignore this suggestion or (2) defer to a followup)

File: r/R/record-batch.R
Comment (nealrichardson) [inline r/R/record-batch.R:166]: FWIW there is a base R function `lengths()` that does this (though I don't recall what version it was added in)

File: r/R/record-batch.R
Comment (nealrichardson) [inline r/R/record-batch.R:173]: Call me 👴 but I personally find 

```
arrays[arr_lens == 1] <- lapply(arrays[arr_lens == 1], repeat_value_as_array, max_array_len)
```

easier to read than `modify2(...)`.

File: r/R/util.R
Comment (nealrichardson) [inline r/R/util.R:123]: If you're checking length again here (unnecessary IMO since you're only calling this in the case where you've validated length == 1 already), you could simplify your `modify2()` wrapper and just `map` over all `arrays`, and in here only do the recycling if length == 1.

File: r/tests/testthat/test-RecordBatch.R
Comment (nealrichardson) [inline r/tests/testthat/test-RecordBatch.R:443]: Why does this error, and should it?

File: r/R/record-batch.R
Comment (thisisnic) [reply r/R/record-batch.R:173]: I don't disagree, will update

File: r/R/record-batch.R
Comment (thisisnic) [reply r/R/record-batch.R:166]: Apparently 3.2.0.  I think in our test-r-versions job, we only got back to 3.3.0 and I can't think of anywhere else that we go back to 3.2.0 or before so will update.

Comment (nealrichardson) [reply r/R/record-batch.R:166]: Per DESCRIPTION we require R >= 3.3 (because we depend on packages that require R >= 3.3)

**Comment by ElenaHenderson** [comment]
@ursabot please benchmark name=dataframe-to-table lang=R

**Comment by ElenaHenderson** [comment]
@ursabot please benchmark name=file-read lang=R

**Comment by ElenaHenderson** [comment]
@ursabot please benchmark name=file-read lang=R

File: r/R/record-batch.R
Comment (thisisnic) [reply r/R/record-batch.R:165]: I opened up a ticket to do this in C++, so I figured probably no point duplicating that effort? Though if this seems like a special case that's better off implemented in the R package's C++ layer rather than the source C++, I can look into it.  See discussion on this ticket, @nealrichardson : https://issues.apache.org/jira/browse/ARROW-12789

File: r/R/util.R
Comment (thisisnic) [reply r/R/util.R:123]: I think I was trying to make this function a bit more generic in case it's useful elsewhere, but you make a good point; I'll remove the check.

File: r/tests/testthat/test-RecordBatch.R
Comment (thisisnic) [reply r/tests/testthat/test-RecordBatch.R:443]: It does it because when we call `length()` on a `tibble`, it returns the number of columns, so in the case in the examples, 2. 

I don't know if it should - should we support recycling for tibbles? I guess we could - I suppose it's consistent with the other behaviour implemented.  This is definitely a problem though as if one of my tibbles has 1 column then weird things happen.

Comment (thisisnic) [reply r/tests/testthat/test-RecordBatch.R:443]: I had a think about this, and `tibble::tibble` does support tibble recycling, so if we're trying to allow similar behaviour, I think it makes sense to implement it here.

Comment (thisisnic) [reply r/tests/testthat/test-RecordBatch.R:443]: @nealrichardson I've now updated this so that the code uses either `length` or `nrow` depending on whether the argument is something that inherits from a `data.frame` or not.  

I got stuck trying to recycle tibbles as I couldn't think how to do it in a reasonable way with the packages we import.

Do you think there is one?  If not, I'm tempted to say it's out of the scope of this ticket anyway.

File: r/R/util.R
Comment (nealrichardson) [inline r/R/util.R:142]: Why not data frames? they turn into struct arrays:

```
> Scalar$create(tibble::tibble(a=1))$as_array()
StructArray
<struct<a: int32>>
-- is_valid: all not null
-- child 0 type: int32
  [
    1
  ]
``

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: r/R/table.R
Comment (thisisnic) [reply r/R/table.R:175]: I was thinking it's OK that there's no scalar recycling with a grouped_df as scalar recycling only happens when there's >1 input passed in, and grouping only happens when there's 1 input passed in.

Using tibble/dplyr, I get `FALSE` if I call `tibble::tibble(slice(iris,1), group_by(iris, Species), .name_repair = "unique") %>% is_grouped_df()`, so this I think this is what we want?

I've refactored the code so things are more in keeping with the early return style & only 1 Table__from_dots call now.

File: r/R/util.R
Comment (thisisnic) [reply r/R/util.R]: RecordBatch__from_arrays fails with a `StructArray` as input due to `count_fields` (from recordbatch.cpp) checking if the object inherits from a data.frame - I've tried to implement something in `18be171`, but it's wrong, and I'm struggling to figure out how to do it correctly.

Comment (nealrichardson) [reply r/R/util.R]: @romainfrancois can you take a look please?

Comment (thisisnic) [reply r/R/util.R]: I've now ended up removing this extra code to try to handle mixed length tibble/data.frames as it seems like a lot of effort for something that doesn't seem to be a particularly common use case.  I've added in an error message for if the input tables are tibbles of different lengths.

File: r/R/util.R
Comment (nealrichardson) [inline r/R/util.R:169]: You added whitespace throughout the package but not here :)

```suggestion
    if (all(map_lgl(arrays, ~inherits(.x, "data.frame")))) {
```

File: r/R/util.R
Comment (nealrichardson) [inline r/R/util.R:172]: right?

```suggestion
          x = paste("Number of rows in inputs:", oxford_paste(arr_lens)
```

Also, are we worried that `arr_lens` could be large (and thus this error message would be huge)?

**Review by nealrichardson — APPROVED** [review]
Some notes but generally LGTM, thanks

File: r/R/util.R
Comment (thisisnic) [reply r/R/util.R]: I wasn't before, but now I am now you mention it.  I have updated my error message to just print the longest and shortest length items seeing as I think this is still sufficient to be useful.  Let me know if it looks OK!

**Pull request closed without merging**
